ERL-558 Add the missing function clause for string:prefix

SeyedMir opened the pull request:
The `string:prefix` function will result in function clause exception in `string:prefix_1` if the second argument is an empty binary. So, `string:prefix(<<"abc">>, <<>>)` will not match any function clauses. This PR fixes this issue.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/erlang/otp?pullRequest=1702) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/erlang/otp?pullRequest=1702) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/erlang/otp?pullRequest=1702) it.</sub>

ferd commented:
other strings like `[<<>>]` or `[[]]` or `[[<<>>]]` are also technically 0-length strings.

SeyedMir commented:
Thanks for letting me know about this. So, the function clause needs to be added to the `string:prefix_1` function; I updated the PR accordingly and added some more test cases.

dgud commented:
Isn't it better to do the test in prefix/2 function once, now you have added an extra test 
in the "hot loop".

Prefix = unicode:characters_to_list(Prefix0),
Prefix can be checked against '[]' before call to prefix_1(..)

Or am I missing something?

SeyedMir commented:
It cab be checked that way too (adding another `case`?), but are we sure that `prefix_1` is not passed an empty list `[]` as its second argument by any other functions? Because for instance `bin_search_str_1` and some other functions call `prefix_1` too.

dgud commented:
We should be, otherwise we should check that before calling `prefix_1` or call `prefix` instead in those
functions. 

But that is another bug in those functions and not a bug in `prefix`, and if you find
any problems there you are free to fix them in the same PR :-) 
Probably they should there input as well.

SeyedMir commented:
OK, I updated the patch and put the check in the body of the `prefix` function. Is that what you were having in mind?

dgud commented:
Thanks

SeyedMir commented:
Just a quick note that the title of the commit should not have been `Add the missing function clause for string:prefix` anymore. It should be something along the lines of `Check prefix0 against [] in the body of string:prefix to capture all variants of empty strings`

dgud commented:
Yes that was my fault I didn't handle/understand github's squash and merge correct, in the future I should ask you to squash the commits before merging. 
I thought it was the merge commit text which is not as important only that the OTP-#### is included,
but squash and merge also rebases to the top and then a ff-only merge.

I'll do it better next time :-) hopefully.
